metrax_example.ipynb

```diff
@@ -0,0 +1,526 @@
+        "\n",
+        "4.  **Aggregate (Reduce):** To get a final, globally correct result, you call the `.reduce()` method on the output from `pmap`. This performs the necessary cross-device communication (an `all_reduce` sum) to combine the intermediate states from all devices into one.\n",
+        "\n",
+        "---\n",
+        "\n",
+        "### Method 2: The `jit` and `Mesh` Approach (Advanced Parallelism)\n",
+        "\n",
+        "For more advanced control over distributed computation, JAX provides an explicit sharding mechanism using the `jax.sharding` API. This **SPMD (Single Program, Multiple Data)** approach is more powerful and flexible than `pmap` and is the standard for large-scale models.\n",
+        "\n",
+        "Instead of `pmap`'s automatic behavior, you take explicit control over each step of the process:\n",
+        "\n",
+        "1.  **Define a `Mesh`**: You first create a logical grid of your physical devices and give names to the axes (e.g., `Mesh(jax.devices(), ('data',))`). This describes the topology you'll be working with.\n",
+        "\n",
+        "2.  **Create a `Sharding` Rule**: You specify exactly how each dimension of your array should be mapped to the mesh's axes. This is done using `NamedSharding` and `PartitionSpec`. For data parallelism, you would shard the batch axis of your data across the `'data'` axis of your mesh.\n",
+        "\n",
+        "3.  **Explicitly Place Data**: You use `jax.device_put` to apply this sharding rule to your data arrays. At this point, your JAX arrays are \"aware\" of how they are distributed across the physical hardware.\n",
+        "\n",
+        "4.  **`jit`-Compile the Function**: You write a function that looks like a normal, single-device calculation and decorate it with `@jax.jit`. When JAX's compiler sees that the inputs to this function are sharded arrays, it automatically generates a distributed version of the code. It implicitly handles all the necessary cross-device communication, so **no explicit `.reduce()` call is needed.**\n",
+        "\n",
+        "This method provides the fine-grained control that is essential for more complex scenarios like model parallelism (sharding a model's weights) in addition to data parallelism."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "n3r37x3sMQbx"
+      },
+      "outputs": [],
+      "source": [
+        "import jax\n",
+        "import numpy as np\n",
+        "import metrax\n",
+        "from jax.sharding import Mesh, NamedSharding, PartitionSpec\n",
+        "\n",
+        "# This script assumes that the JAX environment is configured for 4 devices\n",
+        "# and that the data arrays `predictions`, `labels`, and `sample_weights`\n",
+        "# have been created in a previous cell.\n",
+        "\n",
+        "# --- 1. Metric Calculation Functions ---\n",
+        "\n",
+        "# Method 1: pmap (Simple Data Parallelism)\n",
```

Comment:
This is great! Just confirming: is pmap still recommended? It seems like maybe shard_map is the new API for manual parallelism[1], though for this intro guide I think it could make sense to only discuss `jit()`.

[1] https://github.com/jax-ml/jax/discussions/20312#discussioncomment-8841269